README.md
# LAPS

___

Web based app (nodejs) for showing lap data coming from [RC HourGlass](https://github.com/mv4wd/RCHourglass "RC HourGlass github project page") DIY lap timing system hardware.

I would like to thank Marco Venturini - mv4wd for the use of RC HourGlass logo - see licensing terms below

[Screenshots and help here](./image/SHOW.md)

## Making it work

___

Check [CHANGELOG.md](./CHANGELOG.md) for what is new and I put helpful info in there sometimes

### Windows PC install

Get [NodeJS](https://nodejs.org/en/)

* Get 12.xx.x LTS version

* Install NodeJS

Download my git repository

* use download button or git clone it

* run `npm install` in the dir the package.json file ends up in. Run this when you update also I may have added a new module to the project.

* run `node laps.js --serialport your com port`  example: `node laps.js --serialport com5`

### Raspberry PI Install

```shell
sudo apt install nodejs build-essential
wget https://github.com/snowblind/LAPS/archive/master.zip
unzip master.zip
cd LAPS-master
npm install
node laps --serialport /dev/ttyACM1
```

#### Find serial port
  
```shell
root@RPI3-MAIN:~/RCLAPS # lsusb
Bus 001 Device 005: ID 04b4:0008 Cypress Semiconductor Corp.

root@RPI3-MAIN:~/RCLAPS # ll /dev/serial/by-id

lrwxrwxrwx 1 root root 13 May  4 20:21 usb-MV_Lap_timing_USB_Lap_Timing -> ../../ttyACM1
```

* full setup coming soon

### Errors

If you get a `Error: listen EADDRINUSE :::80` when trying to start node laps.js it means you already have a web server using port 80. Will make this a command line option to set different port at some point. Note port 80 is the default port for web servers.  

## Config Options

___

[transponders.json:](./transponders.json)

* by default it is filled with all transponder IDs included with the RC HourGlass project

* edit the name field to have a name show up and not just the ID

* I may add a way to edit this in the web app

* This is not required for laps to work

## Project status

___

Right now this is a "personal" type lap software. You can not see everybody elses laps times at the same you you are looking at yours. I dunno if it will ever be a full race type program.

It may look ugly to you and I don't care. It works for what I want. The css, html files are changeable.  

I removed the limit of one connection per transponder page.

## Licensing terms

___

* LAPS:

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANYKIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.

IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE." Michael Pounders - snowblind

* RC HourGlass:

Short version: if you build it for your personal use, you're only kindly requested to donate a small sum to a children
charity of your choice (use 'RC Hourglass for children' as a reference).

Please feedback your donations to:charity dot rchourglass at gmail.com

Please consider donating 5 euros per transponder and 30 euros for the decoder for personal use.
If the transponder is used in a club/circuit with an admission fee, please consider donating 100 euros for the decoder.

The original RCTech thread author/designer (Howard Cano) license applies to the decoder project:

*"All information presented on this thread is free for use for personal, non-commercial purposes.
Contact me for licensing arrangements if you wish to produce and market the decoder."* -user howardcano on rctech

Additional licensing terms for the PSOC firmware/design & transponder by mv4wd:

"All information shared is free for use for personal, non-commercial purposes.
Contact me for licensing arrangements if you wish to produce and market the decoder.

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
Software.

Any derivative work source code/design must be public and use this licensing terms.

Any device derived from this project must respond to the command 'License' with this text.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANYKIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.

IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE." Marco Venturini - mv4wd


CHANGELOG.md
# Changelog

## [Unreleased] (00/00/2020)

* add more stats to transponder page like `average lap`, `best lap`, and `worse lap`


## [V0.0.4] (05/15/2020)

### Added

* Command line usage output now with --help
* Changed the html grid layout on devices with screen <768 wide. Looks better on phones. Updated screenshots.
* Lap area now auto scrolls to the bottom to show newest incoming 

### Changes

* Version now comes from package.json. removed VERSION file

### BUGFIX

* There was a problem with the handling of the option being passed to the serialport start up code that would give a `TypeError: First argument must be a string` error with some uses of --serialport and the default option

## [V0.0.3] (05/11/2020)

* Some code clean up
* Added READY GO info to laps area to show what state you are in. First run over the loop detects the transponder, READY waiting again for you to pass over the loop, GO = lap started. Would you like a default list of transponders already detected on load up to skip detection step. Open a github issue.
* Add back in stats shown on transponder page (just fastest lap for now)
* Added command line options --wwwport  example: `node laps.js --wwwport 8080` linux requires root to open port under 1024. No command line --help text atm

## [V0.0.2] (05/04/2020)

* Removed limits on websocket connection that limited the use of transponder pages to one user
* Full updating of all web pages without reloading it yourself. The home page didn't do that last version
* Added command line options --serialport  example: `node laps.js --serialport com5` No command line --help text atm

## [V0.0.1] (04/29/2020)

* first commit


laps.js
const commandline = require('./util/handle_commandline')

if( commandline.options.help ){ 
  commandline.printUsage()
  process.exit(0)
}

const sprintf = require('sprintf-js').sprintf
const SerialPort = require('serialport')
const Readline = require('@serialport/parser-readline')
const port = new SerialPort( commandline.options.serialport.toString(), { baudRate: 115200 } )
const parser = new Readline()
port.pipe(parser)

const express = require('express')
const app = express()
const expressWs = require('express-ws')(app);
const VERSION = require('./package.json').version

const fs = require('fs');
const home_html_cached = fs.readFileSync( __dirname + '/web/html/home.html').toString();
const about_html_cached = fs.readFileSync( __dirname + '/web/html/About.html').toString();
const transponder_html_cached = fs.readFileSync( __dirname + '/web/html/transponder.html').toString();

var found_transids = [];
var rootclients = new Set();
var tlist = require( "./transponders.json");            //TODO:MJP Add a way to edit this from a web page maybe

var appData ={
  flap: 0,
  flapid: 0,
  tottransfound: 0,
  fastlap: 0, //TODO:MJP This is fastest lap for all transponders seen. Need transponder id ? need name also put in new object inside this one

              //TODO:MJP list of found transponders?
};

var myData = { 
  tid: 0,
  lapcount: 0,
  lastlap:  0,
  firstseen: 0,
  fastlap: 0, 
  name: '',   
  clients: null,
  laps: null    
}; 

//setup transponder id with info for testing
if( commandline.options.debug )
{ 
  found_transids[7531106] = { 
    tid: 7531106,
    lapcount: 3,
    lastlap: 109615405,
    firstseen: 109591806,
    fastlap: 5035,
    name: 'JoeP',
    laps: { lap1: '00:07.748', lap2: '00:10.816', lap3: '00:05.035' } 
  }
  found_transids[7531106].clients = new Set()

  found_transids[4572215] = {
  tid: 4572215,
  lapcount: 3,
  lastlap: 2351809139, 
  firstseen: 2351793536, 
  fastlap: 3744,
  name: '', 
  laps: { lap1: '00:03.744', lap2: '00:07.985', lap3: '00:03.874' }
  }
  found_transids[4572215].clients = new Set()
}
app.use('/web', express.static('web'));
app.use('/scripts', express.static('web/scripts'));
app.use('/icons', express.static( 'web/icons'));


app.ws('/', function(ws, req){
  rootclients.add(ws)    
    ws.on('close' , function() {
      rootclients.delete(ws)
    });    
  updateRootWebPage();
});

    
app.ws("/*", function(ws, req){

    var tid = parseInt( req.path.substr(1, 7) )    //TODO:MJP not really robust
    
    found_transids[tid].clients.add(ws)

    ws.on('close', function(msg) {
      found_transids[tid].clients.delete(ws)
    });

    ws.on('message', function(msg) {
      try{                       
            myJson = JSON.parse(msg);
            //console.log( myJson );

            switch(myJson.command){
              case 22:
                clearUserData( myJson.tid );
              break;
            }
            return
          } catch(err) { console.log( err ) }
    });

  updateTransponderWebPage( tid, { command: 44, tid: tid, name: found_transids[tid].name } );

  if( found_transids[tid].firstseen ){
    updateTransponderWebPage( tid, { command: 66, laps: found_transids[tid].laps } );
  }
  else{
    updateTransponderWebPage( tid, { command: 67 } );
  }  
  updateTransponderWebPage( tid, { command: 80, fastlap: timeTotxt(found_transids[tid].fastlap) } );     
});


app.get('/', (req, res) =>{     // HTML ROOT
  res.send( home_html_cached )          
})

app.get('/About.html', (req, res) =>{     // HTML ABOUT   
  res.send( about_html_cached )           
})

app.get('*', (req, res) =>{    // HTML Handle transponder ids

  var tid = parseInt( req.path.substr(1, 7) )
  var ftid = found_transids[tid]
 
  if( !found_transids[tid] ){
    res.send('transponder ID not seen yet')
  return
  }
  
res.send( transponder_html_cached )

})

if( commandline.options.wwwport == 80 ){                                                                                             //Start Web Server
  app.listen( commandline.options.wwwport.toString(), () => console.log( 'LAPS V' + VERSION + ' web app listening at http://yourserveripadress/' ))                       
} 
else{ 
  app.listen( commandline.options.wwwport.toString(), () => console.log( 'LAPS V' + VERSION + ' web app listening at http://yourserveripadress:' + options.wwwport + '/' ))          
}

parser.on('data', line => {

  //console.log("parser.on:start \n" + "    " + transid ) //debug

        var transid = parseInt( line.substr(0, 6), 16 );                                               //TODO:MJP This checking may need to be more robust              
        
        if(!transid)                                                                                     
          return;
        
        let hrtime = process.hrtime();
        let time = parseInt(((hrtime[0] * 1e3) + (hrtime[1]) * 1e-6));

        if( !found_transids[transid] )                                                                 //Only run when we find a new transponder
        {  
          found_transids[transid] = Object.create( myData ); 
          found_transids[transid].laps = Object.create( Object );                                      //Formatted for web use
          found_transids[transid].clients = new Set()
          found_transids[transid].tid = transid;                
          found_transids[transid].name = findFname(transid)
            
            
            updateRootWebPage();
          
            console.log("New Transponder found: " + transid)
          return
        }
        

        if( !found_transids[transid].firstseen )
        {
          found_transids[transid].lastlap = found_transids[transid].firstseen = time;
          updateTransponderWebPage( transid, { command: 68 } )
          return        
        }  

        updateFastLap( curlap = (time - found_transids[transid].lastlap), found_transids[transid] );
                    
        let clap = sprintf( "lap%i",  ++found_transids[transid].lapcount )  //Make string needed for laps key name and +1 lap count
            
        found_transids[transid].laps[clap] = timeTotxt(curlap); 

        
        found_transids[transid].lastlap = time;

        //console.dir(found_transids[transid], {colors: true, depth: 2 }) //debug
        
        if( found_transids[transid].clients ){
          updateTransponderWebPage( transid, { command: 66, laps: found_transids[transid].laps } )
        }
            

  //console.log("parser.on:end \n" ) //debug
  })

function findFname(tid){
  var t = ''
    tlist.forEach(function(item){
      if( item.id == tid )
          t = item.fname;        
    })
  return t;  
}

function updateFastLap( curlap, user ){
    if( ! user.fastlap || curlap < user.fastlap ){
      user.fastlap = curlap;
      updateTransponderWebPage( user.tid, { command: 80, fastlap: timeTotxt(curlap) } );   
    }
}


function timeTotxt( curtime ){
    
    let ms_count = curtime % 1000;
    let ctime = curtime / 1000;
    let secs_count = ctime % 60;
    let mins_count = ( ctime / 60) % 60;

    let laptxt = sprintf( "%02u:%02u.%03u", mins_count, secs_count, ms_count )
    return laptxt
}

function clearUserData( tid ){
    found_transids[tid].lapcount = 0;
    found_transids[tid].laps = {};
    found_transids[tid].lastlap = 0;
    found_transids[tid].firstseen = 0;
    found_transids[tid].fastlap = 0;

    console.dir(found_transids[tid], {colors: true, depth: 1 }) //debug
}

function updateRootWebPage(){                                 //Not sure if this version is faster or not.. need to test on RPI node 10.17.0
  let body =''
    found_transids.forEach(function(item){       
        if( item.name ){
          body += '<li><a href="' +  item.tid + '" ref=>Transponder: ' + item.name + ':' + item.tid + '</a></li>'
        }
        else{
          body +=  '<li><a href="' +  item.tid + '" ref=>Transponder: ' + item.tid + '</a></li>'      
        }
                               
    });
  rootclients.forEach(client =>
  {
      client.send( JSON.stringify( { command: 10, body: body } ) ); 
  });                                    
}

function updateTransponderWebPage( tid, command){
  found_transids[tid].clients.forEach(client =>
  {
      client.send( JSON.stringify( command ) );
  });
}

web/scripts/root.js

var connection = new WebSocket('ws://'+location.hostname, ['arduino']);

connection.onopen = function (){
    console.log('WebSocket connection opened');    
};

connection.onerror = function (error){
    console.log('WebSocket Error ', error);
};

connection.onmessage = function (e){ 
    try{
        var jsonmsg = JSON.parse( e.data )
          
                switch(jsonmsg.command){
                    case 10:
                      document.getElementById('tlist').innerHTML = jsonmsg.body;
                      console.log( 'command 10' );
                    break;
                }
        }
        catch(err){
            console.log( err )
        }    
    
    console.log(e);
};

connection.onclose = function(){   
    console.log('WebSocket connection closed');
};
function load() {
    console.log("load event detected!");  
}
window.onload = load;

web/scripts/WebSocket.js

var connection = new WebSocket('ws://'+location.hostname + location.pathname, ['arduino']);

connection.onopen = function () {
    
    console.log('WebSocket connection opened');
    
};

connection.onerror = function (error) {
    console.log('WebSocket Error ', error);
};

connection.onmessage = function (e) { 
    var body = '';

    var test = JSON.parse( e.data )

        for (var key in test) {
           body += "<li>" + key + " " + test[key] + "</li>"
        }
        body += "<p></p><p></p>"
        document.getElementById('lapdata').innerHTML = body;

};

connection.onclose = function(){
    console.log('WebSocket connection closed');

};
 
function clearUserData() {

    var appData ={
        command: 22,
        tid: location.pathname.substr(1, 7)   
    }
    connection.send(  JSON.stringify(appData) );

    document.getElementById('lapdata').innerHTML = "";
}

function load() {
	console.log("load event detected!");
   
}
window.onload = load;

//console.log( "WebSock.js: " + location.pathname )

web/scripts/transponder.js

var connection = new WebSocket('ws://'+location.hostname + location.pathname, ['arduino']);

var lapdata;
const scrollIntoViewOptions = { behavior: "smooth", block: "end" };

connection.onopen = function (){  
    console.log('WebSocket connection opened');  
};

connection.onerror = function (error){
    console.log('WebSocket Error ', error);
};

connection.onmessage = function(e){ 
    var body = '';
 try{
    var jsonmsg = JSON.parse( e.data )
      
              switch(jsonmsg.command){
                case 44:
                  console.log( 'command 44' );
                  if( jsonmsg.name){
                    document.getElementById('tinfo').innerHTML = jsonmsg.name + ':' + jsonmsg.tid + '<input type="button" onclick="clearUserData()" style="border: none; height: 24px; width: 24px; background: url( /icons/icons8-cls-24.png )"/>'
                  }
                  else{
                    document.getElementById('tinfo').innerHTML =  jsonmsg.tid + '<input type="button" onclick="clearUserData()" style="border: none; height: 24px; width: 24px; background: url( /icons/icons8-cls-24.png )"/>'      
                  }
                break;
                case 66:
                    for (var key in jsonmsg.laps) {
                        body += '<li>' + key + ' ' + jsonmsg.laps[key] + '</li>'
                     }
                     body += '<p></p><p></p>'
                     lapdata.innerHTML = body;
                     lapdata.scrollIntoView(scrollIntoViewOptions);
                     console.log( 'command 66' )
                break;
                case 67:
                     lapdata.innerHTML = '<h1 style="color:yellow">READY!!</h1> Pass over the loop to start';
                     console.log( 'command 67' )
                break;
                case 68:
                     lapdata.innerHTML = '<h1 style="color:green">GO!!!</h1>';
                     console.log( 'command 68' )
                break;
                case 80:
                     document.getElementById('userstats').innerHTML = '<td class="drt1">Fastest Lap</td><td class="drt2">' + jsonmsg.fastlap + '</td>';
                     console.log( 'command 80' )
                break;
                case 81:
                     body += '<p></p><p></p>'
                     lapdata.innerHTML = body;
                     console.log( 'command 81')
                break;
              }
    }
    catch(err){
        console.log( err )
    }
    console.log(e);
}; 

connection.onclose = function(){
    console.log('WebSocket connection closed');
};
 
function clearUserData(){
    connection.send(  JSON.stringify( { command: 22, tid: location.pathname.substr(1, 7) }) );
    lapdata.innerHTML = '<h1 style="color:yellow">READY!!</h1> Pass over the loop to start';
}

function load(){
  lapdata =  document.getElementById('lapdata');
	  console.log( 'load event detected!' );   
}
window.onload = load;

util/handle_commandline.js
const commandLineArgs = require('command-line-args')
const commandLineUsage = require('command-line-usage')

const optionDefinitions = [
     
    { name: 'serialport', type: String, multiple: false, defaultValue: [ "com5" ], description: 'Which serial port to open. See examples area.' },
    { name: 'wwwport', type: Number, multiple: false, defaultValue: [ 80 ], description: 'Allows you to change server port to 8080 if default port is in use or you can not run as root on linux' },
    { name: 'verbose', alias: 'v', type: Boolean, description: 'Display debug print outs' },
    { name: 'debug', alias: 'd', type: Boolean, description: 'For debugging app' },
    { name: 'help', alias: 'h', type: Boolean, description: 'Display this usage guide.' }
  ]

const options = commandLineArgs(optionDefinitions)
  

const sections = [
    {
      header: 'LAPS',
      content: 'Web based app (nodejs) for showing lap data coming from RC HourGlass DIY lap timing system hardware.'
    },
    {
      header: 'Options',
      hide: [ 'debug' ],
      optionList: optionDefinitions
    },
    {
        header: 'Examples',
        content: [
          {
            desc: '1. win32 serial ',
            example: '$ node laps --serialport com2'
          },
          {
            desc: '2. linux serial ',
            example: '$ node laps --serialport /dev/ttyACM0'
          },
          {
            desc: '3. Find Linux serial port',
            example: '$ ls /dev/serial/by-id'
          }
        ]
    },
    {
      content: 'Project home: {underline https://github.com/snowblind/LAPS}'
    }    
]
const usage = commandLineUsage(sections)

module.exports = {
    options: options,
    printUsage: function () {
        console.log(usage)
      }
};